From d5aae55e3138cd5f27f248f2f5fae56822c9c781 Mon Sep 17 00:00:00 2001
From: Bill Chen <billchen1977@gmail.com>
Date: Sat, 1 Apr 2017 21:05:56 +0800
Subject: [PATCH 64/70] stagefright: Revert Fix SurfaceMediaSource buffer
 search condition when buffer return

 buffer handle stored in MediaBuffer had changed, but the buffer return
code did not change correspondingly, so returned buffer can't be
found in list and lead to crash.

Change-Id: Ifb8231387041f056776727cc54286488ac2f5087
stagefright: Fix buffer handle retrieval in signalBufferReturned
---
 media/libstagefright/SurfaceMediaSource.cpp | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/media/libstagefright/SurfaceMediaSource.cpp b/media/libstagefright/SurfaceMediaSource.cpp
index 75ad2d0dd..976b96002 100644
--- a/media/libstagefright/SurfaceMediaSource.cpp
+++ b/media/libstagefright/SurfaceMediaSource.cpp
@@ -424,9 +424,12 @@ void SurfaceMediaSource::signalBufferReturned(MediaBufferBase *buffer) {
     Mutex::Autolock lock(mMutex);
 
     buffer_handle_t bufferHandle = getMediaBufferHandle(buffer);
+    ANativeWindowBuffer* curNativeHandle = NULL;
 
     for (size_t i = 0; i < mCurrentBuffers.size(); i++) {
-        if (mCurrentBuffers[i]->handle == bufferHandle) {
+        curNativeHandle = mCurrentBuffers[i]->getNativeBuffer();
+        if ((mCurrentBuffers[i]->handle == bufferHandle) ||
+            ((buffer_handle_t)curNativeHandle == bufferHandle)) {
             mCurrentBuffers.removeAt(i);
             foundBuffer = true;
             break;
@@ -442,7 +445,10 @@ void SurfaceMediaSource::signalBufferReturned(MediaBufferBase *buffer) {
             continue;
         }
 
-        if (bufferHandle == mSlots[id].mGraphicBuffer->handle) {
+        curNativeHandle = mSlots[id].mGraphicBuffer->getNativeBuffer();
+
+        if ((bufferHandle == mSlots[id].mGraphicBuffer->handle) ||
+            (bufferHandle == (buffer_handle_t)curNativeHandle)) {
             ALOGV("Slot %d returned, matches handle = %p", id,
                     mSlots[id].mGraphicBuffer->handle);
 
-- 
2.11.0

